File: c/sync-addr.c
/*
 * TODO:
 * search lists?
 * canonicalize hostname
 * Error processing - don't return/exit or bailout
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <netdb.h>
#include <arpa/inet.h>

#include <getdns/getdns.h>
#include <getdns/getdns_extra.h>

/* max possible length of a presentation-format IPv6 string */
#define MAX_IPSTRING 40

void usage(const char *progname)
{
    fprintf(stdout, "Usage: %s [-s46] <hostname1> <hostname2> ...\n"
	    "       -s: only return DNSSEC secured results\n"  
	    "       -4: only return IPv4 addresses\n"
	    "       -6: only return IPv6 addresses\n", progname);
    exit(1);
}


int main(int argc, char * const *argv)
{
    getdns_return_t rc = (getdns_return_t) 0;
    const char     *progname;
    int            opt;
    const char     *hostname;
    uint8_t        secure_only=0, v4_only=0, v6_only=0, exit_status=0;
    uint32_t       status;
    char           ipstring[MAX_IPSTRING];
    getdns_context *context;
    getdns_dict    *response;
    getdns_dict    *extensions = NULL;
    getdns_list    *addresses;
    size_t         cnt_addr;
    getdns_dict    *address;
    getdns_bindata *addr_type;
    getdns_bindata *addr_data;

    if ((progname = strrchr(argv[0], '/')))
	progname++;
    else
	progname = argv[0];

    while ((opt = getopt(argc, argv, "sh46")) != -1) {
	switch(opt) {
	case 's': secure_only = 1; break;
	case '4': v4_only = 1; break;
	case '6': v6_only = 1; break;
	default:  
	    usage(progname);
	}
    }
    argc -= optind;
    argv += optind;

    if (argc <  1) {
	usage(progname);
    }

    while ( (hostname = *argv++) ) {

	if ((rc = getdns_context_create(&context, 1))) {
	    fprintf(stderr, "FAIL: Error creating getdns context: %s\n", getdns_get_errorstr_by_id(rc));
	    return 1;
	}

	if (secure_only) {
	    if (! (extensions = getdns_dict_create())) {
		fprintf(stderr, "FAIL: Error creating extensions dict\n");
		return 1;
	    }
	    if ((rc = getdns_dict_set_int(extensions, "dnssec_return_only_secure", 
					  GETDNS_EXTENSION_TRUE))) {
		fprintf(stderr, "FAIL: Error setting dnssec_return_only_secure: %s\n",
			getdns_get_errorstr_by_id(rc));
		return 1;
	    }
	}


	if ((rc = getdns_address_sync(context, hostname, extensions, &response))) {
	    fprintf(stderr, "FAIL: Error looking up addresses for %s: %s\n", hostname, 
		    getdns_get_errorstr_by_id(rc));
	    exit_status = 1;
	    continue;
	}

	if ((rc = getdns_dict_get_int(response, "status", &status))) {
	    fprintf(stderr, "FAIL: Error obtaining status code, %s\n", getdns_get_errorstr_by_id(rc));
	    exit_status = 1;
	    continue;
	}

	if (status == 903) {
	    fprintf(stderr, "FAIL: %s No secure responses obtained\n", hostname);
	    exit_status = 1;
	    continue;
	} else if (status == 901) {
	    fprintf(stderr, "FAIL: %s Non existent domain name\n", hostname);
	    exit_status = 1;
	    continue;
	}


	if ((rc = getdns_dict_get_list(response, "just_address_answers",
				       &addresses))) {
	    fprintf(stderr, "FAIL: Error getting addresses from response dict: %s\n",
		    getdns_get_errorstr_by_id(rc));
	    exit_status = 1;
	    continue;
	}

	if ((rc = getdns_list_get_length(addresses, &cnt_addr))) {
	    fprintf(stderr, "FAIL: Error getting address lengths list: %s\n",
		    getdns_get_errorstr_by_id(rc));
            exit_status = 1;
            continue;
	}

	if (cnt_addr <= 0) {
	    printf("FAIL: %s: No addresses found.\n", hostname);
            exit_status = 1;
            continue;
	}

	for (size_t i = 0; i < cnt_addr; i++) {
	    
	    if ((rc = getdns_list_get_dict(addresses, i, &address))) {
		fprintf(stderr, "FAIL: Error getting address list: %s\n", getdns_get_errorstr_by_id(rc));
		exit_status = 1;
		break;
	    }

	    if ((rc = getdns_dict_get_bindata(address, "address_type", 
					      &addr_type))) {
		fprintf(stderr, "FAIL: Error getting addr_type: %s\n", getdns_get_errorstr_by_id(rc));
		exit_status = 1;
		break;
	    }

	    if ((rc = getdns_dict_get_bindata(address, "address_data", 
					      &addr_data))) {
		fprintf(stderr, "FAIL: Error getting addr_data: %s\n", getdns_get_errorstr_by_id(rc));
		exit_status = 1;
		break;
	    }

	    if (!strncmp((const char *) addr_type->data, "IPv4", 4)) {
		if (v4_only || (!v4_only && !v6_only))
		    fprintf(stdout, "OK: %s IPv4 %s\n", hostname, 
			    inet_ntop(AF_INET, addr_data->data, ipstring, 
				      sizeof(ipstring)));
	    } else if (!strncmp((const char *) addr_type->data, "IPv6", 4)) {
		if (v6_only || (!v4_only && !v6_only))
		    fprintf(stdout, "OK: %s IPv6 %s\n", hostname,
			    inet_ntop(AF_INET6, addr_data->data, ipstring, 
				      sizeof(ipstring)));
	    } else  {
		/* shouldn't get here */
		fprintf(stderr, "FAIL: Unknown address type\n");
		rc = 1;
		break;
	    }

	}

    }

    return exit_status;
}

File: python/sync-addr.py
#!/usr/bin/env python
#

"""
sync-addr.py: resolve given DNS names into IP addresses. The -s switch
constains answers to only ones secured by DNSSEC. The -4 switch only
returns IPv4 addresses, the -6 switch only IPv6 addresses.

An example run:

    $ sync-addr.py -s www.huque.com www.google.com
    www.huque.com: IPv4  50.116.63.23
    www.huque.com: IPv6  2600:3c03:e000:81::a
    www.google.com: No DNSSEC secured responses found

"""

import getdns, sys, getopt, os.path

def usage():
    progname = os.path.basename(sys.argv[0])
    print("""\
Usage: {0} [-s] [-4|-6] <domain1> <domain2> ...

    -s: only return DNSSEC secured answers
    -4: only return IPv4 address answers
    -6: only return IPv6 address answers

-4 and -6 are mutually exclusive. If both are specified, IPv6 wins.
""".format(progname))    
    sys.exit(1)

try:
    (options, args) = getopt.getopt(sys.argv[1:], 's46')
except getopt.GetoptError:
    usage()
else:
    if not args:
        usage()

extensions = { "return_both_v4_and_v6" : getdns.EXTENSION_TRUE }
desired_addr_type = None

for (opt, optval) in options:
    if opt == "-s":
        extensions["dnssec_return_only_secure"] = getdns.EXTENSION_TRUE
    elif opt == "-4":
        desired_addr_type = "IPv4"
    elif opt == "-6":
        desired_addr_type = "IPv6"

ctx = getdns.Context()

for hostname in args:
    try:
        results = ctx.address(name=hostname, extensions=extensions)
    except (getdns.error, e):
        print(str(e))
        break
    status = results.status
    if status == getdns.RESPSTATUS_GOOD:
        for addr in results.just_address_answers:
            addr_type = addr['address_type']
            addr_data = addr['address_data']
            if (desired_addr_type == None) or (addr_type == desired_addr_type):
                print("%s: %s  %s" % (hostname, addr_type, addr_data))
    elif status == getdns.RESPSTATUS_NO_SECURE_ANSWERS:
        print("%s: No DNSSEC secured responses found" % hostname)
    else:
        print("%s: getdns.address() returned error: %d" % (hostname, status))
